### PR TITLE
[NUI-1324] Add context classes: first draft

### PR DESCRIPTION
**Pull request by tmathern**
## Description

- Adds context classes for the engine, the plan and transformers.
- Use context classes.
The idea is to keep variable states in the context, as well as attributes/properties that may need to be moved along between transformer, or engine, plan and transformers. 

This PR does *not* introduce new fields or behavior yet, but the classes on which we'll most likely add them (eg: instructions executed vs. what is left to do, progress, ...).

Fixes [NUI-1324](https://jira.corp.adobe.com/browse/NUI-1324)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/context/engine-context.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    set directory(value) {
+        this[INTERNAL].directory = value;
+    }
+    
+    get directory() {
+        return this[INTERNAL].directory;
+    }
+}
```
Comment (jdelbick) [inline lib/context/engine-context.js:31]: we may want to add a method or storage object like `currentPlan(s)` that will track the plans currently being run

File: lib/context/plan-context.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    set current(value) {
+        this[INTERNAL].current = value;
+    }
+    
+    get current() {
+        return this[INTERNAL].current;
+    }
+
+    set originalInput(value) {
```
Comment (jdelbick) [inline lib/context/plan-context.js:39]: nice to have: we should make it so you cannot set `originalInput` after its been set (so maybe no setter?)

File: package.json [hunk 3]
```diff
@@ -27,31 +27,31 @@
...
     "clone": "^2.1.2",
     "css-unit-converter": "^1.1.2",
     "data-uri-to-buffer": "^3.0.1",
-    "debug": "^4.3.1",
-    "file-type": "^16.5.0",
+    "debug": "^4.3.2",
+    "file-type": "^16.5.3",
     "fs-extra": "^10.0.0",
-    "graph-data-structure": "^1.15.0",
+    "graph-data-structure": "^2.0.0",
```
Comment (jdelbick) [inline package.json:54]: I wouldn't update the major version of this inside this PR

File: lib/context/transformer-context.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    get plan() {
+        return this[INTERNAL].step;
+    }
+
+    get step() {
+        return this[INTERNAL].step;
+    }
+
+    get transformer() {
```
Comment (jdelbick) [inline lib/context/transformer-context.js:106]: we may also want a `transformerName` getter

**Review by jdelbick — COMMENTED** [review]
LGTM,

I like the engine context and transformer context. they are simple but effective and will help simplify the code in `engine.js`

However, could you explain your reasoning for adding a plan context?

File: package.json (hunk 3 above)
Comment (tmathern) [reply package.json:54]: Why? We have no breaking changes to account for. Any concerns?

File: lib/context/engine-context.js (hunk 1 above)
Comment (tmathern) [reply lib/context/engine-context.js:31]: I didn't add it on purpose, because for now we don't need it. (I'd expect the need to appear in the future).

**Pull request closed without merging**